libretro-common/rest/rest.c

```diff
@@ -0,0 +1,202 @@
+{
+   struct rest_request_t *request;
+
+   request = (rest_request_t *)calloc(1, sizeof(rest_request_t));
+   request->http_request = http_request;
+
+   return request;
+}
+
+void rest_request_set_header(rest_request_t *request, const char *name, const char *value, bool replace)
+{
+   net_http_request_set_header(request->http_request, name, value, replace);
+}
+
+void rest_request_set_retry_policy(
+   rest_request_t *request,
+   rest_retry_policy_t *retry_policy)
+{
+   if (request)
+   {
+      request->retry_policy = retry_policy;
+   }
+}
+
+static void _start_request(rest_request_t *request)
+{
+   request->http_conn = net_http_connection_new(request->http_request);
+   if (!request->http_conn || !net_http_connection_iterate(request->http_conn) || !net_http_connection_done(request->http_conn))
+   {
+      net_http_connection_free(request->http_conn, true);
+      request->http_conn = NULL;
+      return;
+   }
+
+   request->http = net_http_new(request->http_conn);
+   if (!request->http)
+   {
+      net_http_connection_free(request->http_conn, true);
+      net_http_delete(request->http);
+      request->http_conn = NULL;
+      request->http = NULL;
+   }
+}
+
+struct http_response_t *rest_request_execute(rest_request_t *request)
+{
+   time_t timeout;
+   size_t progress;
+   size_t total;
+
+   timeout = time(NULL) + 30;
+
+   do {
+      if (!request->http_conn)
+      {
+         _start_request(request);
+      }
+
+      if (net_http_update(request->http, &progress, &total))
+      {
+         struct http_response_t *response;
+
+         response = net_http_get_response(request->http);
+
+         if (request->retry_policy)
+         {
+            int status_code;
+            int i;
+            bool should_retry = false;
+            struct rest_api_response_callback_with_status_t *current_entry;
+
+            status_code = net_http_response_get_status(response);
+
+            for (i = 0;i < request->retry_policy->status_codes_len;i++)
+            {
+               if (status_code == request->retry_policy->status_codes[i])
+               {
+                  should_retry = true;
+                  break;
+               }
+            }
+
+            if (should_retry)
+            {
+               net_http_delete(request->http);
+               net_http_connection_free(request->http_conn, false);
+               request->http = NULL;
+               request->http_conn = NULL;
+
+               if (request->retry_count < request->retry_policy->delays_len - 1)
+               {
+                  request->delay_end = time(NULL) + request->retry_policy->delays[request->retry_count];
+                  request->retry_count++;
+               } else
+               {
+                  return NULL;
+               }
+            } else
+            {
+               return response;
+            }
+         } else
+         {
+            return response;
+         }
+      } else
+      {
+         usleep(100);
```

Comment:
just break here?

Reply:
The idea is to go back to the top of the loop and try reading more HTTP data from the connection.

Reply:
Ah had to look at the file in the IDE, thought it was the else of request->retry_policy